### DISCUSSION TITLE
Retrieval Experiments with all-MiniLM-L6-v2 + simcse

### DISCUSSION THREAD

**Discussion by yyLeaves** [Kaggle-Competitions-Code/k12-curriculum-recommendations#5] [General]
[exp data](https://docs.google.com/spreadsheets/d/1kcMuXYRLzUOY2_acmWKA8ek-WdM7ZZ55-jhMh9XkZU8/edit?usp=sharing)

shuffled - 指groupby language然后shuffle

## 1. base setups
```
batch_size = 64
max_len = 64
dataset not shuffled
```
Results
|top_n|overall recall|avg line recall|
|-|-|-|
|10|0.4096557560858269|0.4917691086703409|
|20|0.513151959718295|0.5764448464903298|
|50|0.6403634152790062|0.6740202915667366|
|100|0.7265405096219296|0.7441767731548785|
|200|0.8042631897142054|0.8091437845276167|

## 2. increase batch size from 64 -> 128 
**1->2**
```
batch_size = 128
max_len = 64
dataset not shuffled
```
Results
|top_n|overall recall|avg line recall|
|-|-|-|
|10|0.4324056152054837|0.515438373956131|
|20|0.5421514208567578|0.603538013905496|
|50|0.6738983053889892|0.7051779295701075|
|100|0.7580769189388107|0.7765029684217076 |
|200|0.8339243254251469|0.8460831447860021|

## 3. group dataset by language then shuffle
**1->3**
```
batch_size = 64
max_len = 64
dataset shuffled
```
Results
|top_n|overall recall|avg line recall|
|-|-|-|
|10|0.41372134073106254|0.497197837970398|
|20|0.5189953823978796|0.5803661474981462|
|50|0.649738594182533|0.682701312182591|
|100|0.7349298285518711|0.7509243328572374 |
|200|0.812963843997365|0.8167082416526877|

## 4. increase max_len from 64 -> 128
**3->4**
```
batch_size = 64
max_len = 128
dataset shuffled
```
Results
|top_n|overall recall|avg line recall|
|-|-|-|
|10|0.42784422639709857|0.510646598244633|
|20|0.5380966348498997|0.6001088594405399|
|50|0.6633448360080142|0.6982745467193237|
|100|0.7510185691184612 |0.7693480879148443|
|200|0.8281491662824991 |0.8354691332643241|

15个epoch模型还没有完全收敛，继续训练下去结果应该还会有提升

## 5.increase max_len -> 256
**1/3->5**
```
batch_size = 64
max_len = 256
dataset shuffled
continuing from 1's checkpoint Epoch 34668
epoch 19/20 epochs in total
```
初始lr设高了，第19个epoch的效果最好，可能继续训练下去还能有一点点提升
Results
| top_n | overall recall     | avg line recall    | f2 cv               | lb    |
| ----- | ------------------ | ------------------ | ------------------- | ----- |
| 5     | 0.3322644549870804 | 0.4336801549590748 | 0.346316172690926   | 0.362 |
| 10    | 0.4484871192918549 | 0.5330171220384055 | 0.3360662492085829  | /     |
| 20    | 0.5607020272932969 | 0.6197723525222764 | 0.28420618255836644 | /     |
| 50    | 0.6915494691864652 | 0.7203332470018804 | 0.18885967397124015 | /     |
| 100   | 0.7760107527862308 | 0.7945262938553538 | 0.12337644482373789 | /     |
| 200   | 0.8513299666435534 | 0.8642198458294773 | 0.07452541661167635 | /     |

## 8. training simcse with negative samples
**1->8**
```
batch_size = 64
max_len = 64
dataset shuffled
continuing from 1's checkpoint Epoch 34668
hard_negative_weight = 0.05
```
用1的模型召回top100，为每个正样本对sample一个负样本
| top_n | overall recall     | avg line recall    | f2 cv               | lb    |
| ----- | ------------------ | ------------------ | ------------------- | ----- |
| 5     | 0.3602309964370354 | 0.4657223581451884 | 0.3733758187805152| 0.382 |
| 10    | 0.4807232836091307 | 0.562805166692416| 0.35713529291418367| /     |
| 20    | 0.5874685727373982 | 0.6437847851452513 | 0.29644236838328614| /     |
| 50    | 0.6975022334844584 | 0.7295144467700183 | 0.19110383784437263| /     |
| 100   | 0.7659645069282203 | 0.7849919386005749 | 0.12337644482373789 | /     |

## 9. scale up training simcse with negative samples
**8->9**
```
batch_size = 192
max_len = 128
dataset shuffled
continuing from 2's checkpoint Epoch 28248
hard_negative_weight = 0.05
```

| top_n | overall recall     | avg line recall    | f2 cv               | lb    |
| ----- | ------------------ | ------------------ | ------------------- | ----- |
| 5     | 0.371721771685636 | 0.4790336690175793 | 0.3840771380204225| 0.41 |
| 10    | 0.49829423833940967 | 0.5808105814609091 | 0.3687065550833983| /     |
| 20    | 0.6104513331492541 | 0.6645875823982077| 0.3065033739313819| /     |
| 50    | 0.7237736033772875 | 0.7560558489295186| 0.19717597024196634| /     |
| 100   | 0.7933331886876476 | 0.8165777703677591  | 0.12566782048874725| /     |

## 10. boost model with fresh negative samples from top75
**9->10**
```
batch_size = 192
max_len = 128
dataset shuffled
continuing from 9's checkpoint Epoch 32742
hard_negative_weight = 0.05
nagative samples from top 75 results of 9's model
```
用9的模型召回top75，为每个正样本对sample一个负样本
| top_n | overall recall     | avg line recall    | f2 cv               | lb    |
| ----- | ------------------ | ------------------ | ------------------- | ----- |
| 5     | 0.3917784468383733 | 0.5043960078553491 | 0.4038875261814184| 0.428 |
| 10    | 0.523240748387827| 0.6093793834925005 | 0.385824133813483| /     |
| 20    | 0.6337613677705383  | 0.6937948727796245 | 0.31724050367167167| /     |
| 50    | 0.7420661770355433 | 0.78308396607833 | 0.200905730428947| /     |
| 100   | 0.8089301719107014 | 0.840522967450222 | 0.12731365637676395| /     |

## 11. boost model with fresh negative samples from top50
**10->11**
```
batch_size = 192
max_len = 128
dataset shuffled
continuing from 9's checkpoint Epoch 30816
hard_negative_weight = 0.05
nagative samples from top 50  results of 9's model
```
用10的模型召回top50，为每个正样本对sample一个负样本
| top_n | overall recall     | avg line recall    | f2 cv               | lb    |
| ----- | ------------------ | ------------------ | ------------------- | ----- |
| 5     | 0.40496463138866956 | 0.5207539277936671 | 0.41683166817195816| 0.439 |
| 10    | 0.5389921939415862 | 0.6285949597117355 | 0.39664707240754443| /     |
| 20    | 0.6447428100137511 | 0.7120661503574027 | 0.3222912903774863| /     |
| 50    | 0.7453391308302046 | 0.7948331470661236 | 0.20152617345208518| /     |
| 100   | 0.8088177927914856 | 0.8481638087539346 | 0.12731365637676395| /     |

## 12. boost model with fresh negative samples from top25
**11->12**
用11的模型召回top25，为每个正样本对sample一个负样本
| top_n | overall recall     | avg line recall    | f2 cv               | lb    |
| ----- | ------------------ | ------------------ | ------------------- | ----- |
| 5     | 0.414288603170375 | 0.5352490260647714 | 0.42649016927616806| 0.447 |
| 10    | 0.5447988959694628 | 0.6400883460588352 | 0.40086046440971274| /     |
| 20    | 0.6455033836128192 | 0.7227569531227362 | 0.32285524862512477| /     |
| 50    | 0.7427517503318051 | 0.8011386456614168 | 0.20093415922537403| /     |
| 100   | 0.8050305431265959 | 0.8513663041110231 | 0.12651275281246746| /     |

## CV/LB
CV：exclude category=source in validation set, and use it in all train sets
```python
cv = StratifiedGroupKFold(n_splits=3, random_state=cfg.seed, shuffle=True)
for i, (train_index, test_index) in enumerate(cv.split(df_candidates, df_candidates['language'], df_candidates['id'],)):
        df_candidates.loc[test_index, 'fold'] = i
```
|trail|notes|cv|lb|
|-|-|-|-|
|1|r1.2 Epoch 32742 top50||0.202|
|2|r1.2 Epoch 32742 top5||0.344|
|3|r1.2 Epoch 32742 top10||0.341|
|4|r1 Epoch 34668 top5|0.32076433469655813|0.326|
|5|r1.5 Epoch 73188 top10|0.3360662492085829|0.362|
|8|r1.8 Epoch 65484 top10|0.35713529291418367|0.382|
|8|r1.8 Epoch 65484 top5|0.3733758187805152|0.392|
|9|r1.9 Epoch 28248 top5| 0.3840771380204225|0.41|
|10|r1.10 Epoch 30816 top5|0.4038875261814184|0.428|
|11|r1.11 Epoch 33384 top5|0.41683166817195816|0.439|
|12|r1.12 Epoch 38520 top5|0.42649016927616806|0.447|

**Comment by yyLeaves** [comment]
# 单模型ideas
1. increase max_len from 128 -> 256/512 (资源显存不够)
2. increase batch_size from 192 -> 256/512 (资源显存不够)
3. 增加正负样本比from 1:1 -> 1:2/1:3
4. <strike>tune hard_negative_weight</strike>

**Comment by yyLeaves** [comment]
## Boosting new fold 1 with no source topics

f0r3

| top_n | overall recall     | avg line recall    | f2 cv               | lb    |
| ----- | ------------------ | ------------------ | ------------------- | ----- |
| 5     | 0.4015044275792736 | 0.5220956435923714 | 0.41700373858387574 | / |
| 10    | 0.5448438379520066 | 0.6429301142534748 | 0.4039608611977723  | /     |
| 20    | 0.6627457853736283 | 0.7342118242370695 | 0.33243898467028105 | /     |
| 50    | 0.7720143057068651 | 0.819761398316757  | 0.20924274637534485 | /     |
| 100   | 0.8358561002881366 | 0.8681389735993651 | 0.13205861108960237 | /     |

f0r4

| top_n | overall recall      | avg line recall    | f2 cv               | lb   |
| ----- | ------------------- | ------------------ | ------------------- | ---- |
| 5     | 0.40460083471475794 | 0.5325069107955293 | 0.42247935874837395 | /    |
| 10    | 0.5478970558263647  | 0.6529397544901132 | 0.4066180403799889  | /    |
| 20    | 0.6620970102385854  | 0.7424966167388147 | 0.3326628412074118  | /    |
| 50    | 0.7711616480259168  | 0.8260519337069118 | 0.2090889650034113  | /    |
| 100   | 0.8335869009600388  | 0.8700068976719416 | 0.13164848585585218 | /    |

f1r3

| top_n | overall recall      | avg line recall    | f2 cv               | lb   |
| ----- | ------------------- | ------------------ | ------------------- | ---- |
| 5     | 0.38848999548783264 | 0.5031008472485843 | 0.40360950765153175 | /    |
| 10    | 0.5251884786796032  | 0.6151343941705735 | 0.390001869488584   | /    |
| 20    | 0.6407001184402903  | 0.7046195796049964 | 0.32233586578196516 | /    |
| 50    | 0.7528645946473896  | 0.7954317559788239 | 0.20465115212846335 | /    |
| 100   | 0.8228359825017433  | 0.8513534419193375 | 0.1302313847323367  | /    |

f1r4

| top_n | overall recall      | avg line recall    | f2 cv               | lb   |
| ----- | ------------------- | ------------------ | ------------------- | ---- |
| 5     | 0.40228445998870965 | 0.525970869185446  | 0.418596712953492   | /    |
| 10    | 0.543971462730194   | 0.6477995382768781 | 0.40479820749219536 | /    |
| 20    | 0.6613456801525115  | 0.7380561976208723 | 0.33197188911375797 | /    |
| 50    | 0.7714022345757998  | 0.8227822961019742 | 0.20885433971257916 | /    |
| 100   | 0.8353542115737941  | 0.8684936754277087 | 0.13185689371517986 | /    |

f1r5

| top_n | overall recall     | avg line recall    | f2 cv               | lb   |
| ----- | ------------------ | ------------------ | ------------------- | ---- |
| 5     | 0.4081046437191723 | 0.5376825003389406 | 0.4256060785194115  | /    |
| 10    | 0.5509393615098321 | 0.6594290301733581 | 0.40958744590878793 | /    |
| 20    | 0.6661828624599033 | 0.7470690797182201 | 0.33449581998074207 | /    |
| 50    | 0.7744026309262033 | 0.8262809947095721 | 0.20959160786161002 | /    |
| 100   | 0.8361253597845767 | 0.8697768430625878 | 0.1320088005134462  | /    |

f2r4

| top_n | overall recall     | avg line recall    | f2 cv               | lb   |
| ----- | ------------------ | ------------------ | ------------------- | ---- |
| 5     | 0.4011903940126507 | 0.523293811897878  | 0.41751486725791537 | /    |
| 10    | 0.5455833429486767 | 0.6426282865182675 | 0.4041557280693036  | /    |
| 20    | 0.6606622909434813 | 0.7329684558629167 | 0.33143043564589053 | /    |
| 50    | 0.7727004497108851 | 0.8202786919134787 | 0.20914279852109616 | /    |
| 100   | 0.8356836726232459 | 0.866971278451776  | 0.13205454684701048 | /    |

f2r5

| top_n | overall recall     | avg line recall    | f2 cv               | lb   |
| ----- | ------------------ | ------------------ | ------------------- | ---- |
| 5     | 0.4062979506064163 | 0.5325588256039375 | 0.423561144505964   | /    |
| 10    | 0.5482111416663308 | 0.6529746660195659 | 0.40734273304108354 | /    |
| 20    | 0.6651113843739047 | 0.7402508201709324 | 0.3333535756382739  | /    |
| 50    | 0.776083255994895  | 0.8248152436912071 | 0.2099176359435621  | /    |
| 100   | 0.8407137994854922 | 0.871229412765091  | 0.13263008614030952 | /    |



f3r4

| top_n | overall recall     | avg line recall    | f2 cv               | lb   |
| ----- | ------------------ | ------------------ | ------------------- | ---- |
| 5     | 0.3993512856620332 | 0.516491294888085  | 0.4138492314800401  | /    |
| 10    | 0.5425224904884747 | 0.6360527904530525 | 0.40154636843195796 | /    |
| 20    | 0.6571180731988433 | 0.7250668258816448 | 0.32961941856569327 | /    |
| 50    | 0.7698327684625121 | 0.8124866266224134 | 0.20856985554908852 | /    |
| 100   | 0.8343822843151479 | 0.8627463219508358 | 0.13191100278232432 | /    |


3 fold ensemble

    '/root/autodl-tmp/model/f0r5/checkpoint-6420',
    '/root/autodl-tmp/model/f1r6/checkpoint-5457',
    '/root/autodl-tmp/model/f2r5/checkpoint-4160'




| top_retrieval | head | avg line recall    | f2 cv               | lb   |
| ------------- | ---- | ------------------ | ------------------- | ---- |
| 100           | 5    | 0.5467866494303824 | 0.43449101571885196 | /    |
| 100           | 10   | 0.6702250377553186 | 0.4184049636883239  | /    |
| 100           | 20   | 0.7568919719613939 | 0.3406261993671188  | /    |
| 100           | 50   | 0.8415280722641862 | 0.21332692763061714 | /    |
| 100           | 100  | 0.8892379556163074 | 0.13454268520271054 | /    |
|               |      |                    |                     |      |
| 50            | 50   | 0.8480637324521544 | 0.21407147757395506 |      |
| 50            | 20   | 0.7593033842395931 | 0.34069641627314806 |      |
| 50            | 10   | 0.6704823735508082 | 0.41839694785145376 |      |
| 50            | 5    | 0.5522348169548069 | 0.4376348986098108  |      |
|               |      |                    |                     |      |
| 20            | 5    | 0.548627335030083  | 0.43501223717684717 |      |
| 20            | 10   | 0.676292210454046  | 0.42066691449560595 |      |
| 20            | 20   | 0.7687140488400852 | 0.3429355025487618  |      |
|               |      |                    |                     |      |
| 10            | 10   | 0.6737559951197932 | 0.4194090299131335  |      |
| 10            | 5    | 0.549640214921909  | 0.4351853181628566  |      |
|               |      |                    |                     |      |
| 5             | 5    | 0.5510184190357351 | 0.43618270298951284 |      |

**Reply by EeyoreLee** [reply]
fnrm代表什么

**Reply by yyLeaves** [reply]
> fnrm代表什么

f是fold r是第几轮(这里的r实际要-1)
